dnssec.c

```diff
@@ -409,8 +413,111 @@ ldns_key_buf2dsa_raw(const unsigned char* key, size_t len)
 #endif /* OPENSSL_VERSION_NUMBER */
 	return dsa;
 }
+#endif /* OPENSSL_NO_DEPRECATED_3_0 */
+
+#if OPENSSL_VERSION_NUMBER >= 0x30000000L
+
+EVP_PKEY *
+ldns_dsa2pkey(const ldns_buffer *key)
+{
+	return ldns_dsa2pkey_raw((const unsigned char*)ldns_buffer_begin(key),
+						   ldns_buffer_position(key));
+}
+
+EVP_PKEY *
+ldns_dsa2pkey_raw(const unsigned char* key, size_t len)
+{
+	uint8_t T;
+	uint16_t length;
+	uint16_t offset;
+	EVP_PKEY_CTX *ctx;
+	OSSL_PARAM params[5];
+	EVP_PKEY *pkey = NULL;
+	unsigned char *Q = NULL;
+	unsigned char *P = NULL;
+	unsigned char *G = NULL;
+	unsigned char *Y = NULL;
+
+	if(len == 0)
+		return NULL;
+	T = (uint8_t)key[0];
+	length = (64 + T * 8);
+	offset = 1;
+
+	if (T > 8) {
+		return NULL;
+	}
+	if(len < (size_t)1 + SHA_DIGEST_LENGTH + 3*length)
+		return NULL;
+
+	Q = OPENSSL_zalloc(SHA_DIGEST_LENGTH);
+	P = OPENSSL_zalloc(length);
+	G = OPENSSL_zalloc(length);
+	Y = OPENSSL_zalloc(length);
+
+	if (!Q || !P || !G || !Y) {
+		OPENSSL_clear_free(Q, SHA_DIGEST_LENGTH);
+		OPENSSL_clear_free(P, length);
+		OPENSSL_clear_free(G, length);
+		OPENSSL_clear_free(Y, length);
+	}
+
+	memcpy(Q, key+offset, SHA_DIGEST_LENGTH);
+	ldns_swap_bytes(Q, SHA_DIGEST_LENGTH);
+	offset += SHA_DIGEST_LENGTH;
+
+	memcpy(P, key+offset, length);
+	ldns_swap_bytes(P, length);
+	offset += length;
+
+	memcpy(G, key+offset, length);
+	ldns_swap_bytes(G, length);
+	offset += length;
+
+	memcpy(Y, key+offset, length);
+	ldns_swap_bytes(Y, length);
+
+	/* Exponent */
+
+	params[0] =
+		OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_FFC_Q, Q, SHA_DIGEST_LENGTH);
+
+	params[1] = OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_FFC_P, P, length);
```

Comment:
This may have used ``OSSL_PARAM_set_BN`` with combination of ``BN_bin2bn`` IMO. Is there a reason it were not used for it and own swap implemented?

Reply:
Yes. This is an unfortunate requirement for OpenSSL3, but all OSSL_PARAMs must be stored in *native* byte order instead of network byte order. BIGNUMs created by BN_bin2bn and similar functions store the data in network byte order. This effectively makes OSSL_PARAM_set_BN difficult to use on little-endian systems.

I suppose it would be possible to call BN_lebin2bn to create a backwards BIGNUM on little endian systems, but it's a bit strange because you're calling a function designed to take little endian data on big endian data. It's also weird because you'd have to conditionally call the function based on the endianness of the system.

However, I recall trying this when doing work for libssh2 and finding that it didn't work properly. The only solution that worked for me at the time was swapping the buffer and calling OSSL_PARAM_construct_BN.